monitor/internal/pod/monitor.go

```diff
@@ -196,47 +171,101 @@ func (m *PodMonitor) Resync(ctx context.Context) error {
 	return ResyncWithAllPods(ctx, m.kubeClient, m.resyncInfo, m.eventsCh, m.localNode)
 }
 
-func (m *PodMonitor) startManager(ctx context.Context, mgr manager.Manager) error {
-	controllerStarted := make(chan struct{})
-	if err := mgr.Add(&runnable{ch: controllerStarted}); err != nil {
-		return fmt.Errorf("pod: %s", err.Error())
-	}
+const (
+	startupWarningMessage = "pod: the Kubernetes controller did not start within the last 5s. Waiting..."
+)
+
+var (
+	retrySleep          = time.Second * 3
+	warningMessageSleep = time.Second * 5
+	managerNew          = manager.New
+)
+
+func (m *PodMonitor) startManager(ctx context.Context) {
+	var mgr manager.Manager
 
-	// starting the manager is a bit awkward:
-	// - it does not use contexts
-	// - we pass in a fake signal handler channel
-	// - we start another go routine which waits for the context to be cancelled
-	//   and closes that channel if that is the case
 	startTimestamp := time.Now()
 	z := make(chan struct{})
-	errCh := make(chan error, 2)
-	go func() {
-		<-ctx.Done()
-		close(z)
-		errCh <- ctx.Err()
-	}()
+	controllerStarted := make(chan struct{})
+
 	go func() {
-		if err := mgr.Start(z); err != nil {
-			errCh <- err
+		// manager.New already contacts the Kubernetes API
+		for {
+			var err error
+			mgr, err = managerNew(m.kubeCfg, manager.Options{})
+			if err != nil {
+				zap.L().Error("pod: new manager instantiation failed. Retrying in 3s...", zap.Error(err))
+				time.Sleep(retrySleep)
+				continue
+			}
+			break
+		}
+
+		// Create the delete event controller first
+		dc := NewDeleteController(mgr.GetClient(), m.localNode, m.handlers, m.sandboxExtractor, m.eventsCh)
+		for {
+			if err := mgr.Add(dc); err != nil {
+				zap.L().Error("pod: adding delete controller failed. Retrying in 3s...", zap.Error(err))
+				time.Sleep(retrySleep)
+				continue
+			}
+			break
+		}
+
+		// Create the main controller for the monitor
+		for {
+			if err := addController(
+				mgr,
+				newReconciler(mgr, m.handlers, m.metadataExtractor, m.netclsProgrammer, m.sandboxExtractor, m.localNode, m.enableHostPods, dc.GetDeleteCh(), dc.GetReconcileCh(), m.resyncInfo),
+				m.workers,
+				m.eventsCh,
+			); err != nil {
+				zap.L().Error("pod: adding main monitor controller failed. Retrying in 3s...", zap.Error(err))
+				time.Sleep(retrySleep)
+				continue
+			}
+			break
+		}
+
+		for {
+			if err := mgr.Add(&runnable{ch: controllerStarted}); err != nil {
+				zap.L().Error("pod: adding side controller failed. Retrying in 3s...", zap.Error(err))
+				time.Sleep(retrySleep)
+				continue
+			}
+			break
+		}
+
+		// starting the manager is a bit awkward:
+		// - it does not use contexts
+		// - we pass in a fake signal handler channel
+		// - we start another go routine which waits for the context to be cancelled
+		//   and closes that channel if that is the case
+
+		for {
```

Comment:
So the monitor is going to keep on trying till the controllers start else the enforcer will fail to start. 
That means without the controllers we don't start the enforcer?

Reply:
yes